### PR TITLE
[NP-4929] mlang changes to support fee expressions.

### PR DESCRIPTION
**Pull request by jlhughes**

### CODE REVIEW

File: src/foam/mlang/mlang.js [hunk 1]
```diff
@@ -4501,6 +4501,15 @@ foam.CLASS({
   ],
 
   properties: [
+    {
+      class: 'String',
+      name: 'id',
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js]: Why is this needed?

File: src/foam/mlang/mlang.js [hunk 2]
```diff
@@ -4702,6 +4711,54 @@ foam.CLASS({
...
+    }
+  ],
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: `
+        if ( getPredicate().f(obj) ) {
+          return  getValueIfTrue() != null ?  getValueIfTrue().f(obj) : null;
+        } else {
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js]: No need for the 'else' if you've already returned in the 'if'.

File: src/foam/mlang/mlang.js [hunk 3]
```diff
@@ -4702,6 +4711,54 @@ foam.CLASS({
...
+        }
+      `
+    },
+    {
+      name: 'toString',
+      type: 'String',
+      javaCode: `
+        var sb = new StringBuilder();
+        sb.append("If(")
+          .append("predicate:").append(String.valueOf(getPredicate()))
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js]: String.valueOf() is already handled by overloaded append() methods. Also, since you can do this all in one statement, it's the exact same as you had just done return "If(predicate:" + getPredicate + ", valueIfTrue:" + ...

File: src/foam/mlang/mlang.js (hunk 1 above)
Comment (lchanmann) [reply src/foam/mlang/mlang.js]: I suspect it might not be needed https://github.com/nanoPayinc/NANOPAY/pull/14146#discussion_r818604591.

File: src/foam/nanos/auth/LastModifiedAware.js [hunk 4]
```diff
@@ -13,7 +13,10 @@ foam.INTERFACE({
       class: 'DateTime',
       name: 'lastModified',
       documentation: 'Last modified date',
-      storageOptional: true
+      storageOptional: true,
+      javaCompare: 'return 0;',
```
Comment (lchanmann) [inline src/foam/nanos/auth/LastModifiedAware.js]: Will this still allow for sort/orderBy lastModified?

File: src/foam/mlang/mlang.js [hunk 5]
```diff
@@ -1862,7 +1864,7 @@ foam.CLASS({
       return typeof x === 'number' ? '' + x :
         typeof x === 'string' ? '"' + x + '"' :
         Array.isArray(x) ? '[' + x.map(this.toString_.bind(this)).join(', ') + ']' :
-        x && (x).toString();
+        x && x.toString && x.toString();
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js]: BTW, you can now do this in JS:
x && x.toString?.()

File: src/foam/mlang/mlang.js [hunk 6]
```diff
@@ -4620,7 +4627,8 @@ foam.CLASS({
     {
       name: 'f',
       javaCode: `
-        Double result = null;
+      Double result = null;
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js:4566]: Why is this a Double instead of a double?

File: src/foam/mlang/mlang.js [hunk 7]
```diff
@@ -4807,6 +4836,64 @@ foam.CLASS({
...
+      class: 'foam.mlang.ExprProperty',
+      name: 'falseExpr'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: `
+        if ( getTrueExpr() == null ||
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js]: Maybe switch || to &&. It seems legitimate that one would be null.

Comment (jlhughes) [reply src/foam/mlang/mlang.js]: This is only for logging output.   Most of this logging throughout will be removed before any of this is merged.

File: src/foam/mlang/mlang.js (hunk 6 above)
Comment (jlhughes) [reply src/foam/mlang/mlang.js:4566]: Testing Object for null simplifies the test for first assignment.  
Changing to native double will require a flag for first assignment.

File: src/foam/core/ImportsExports.js [hunk 8]
```diff
@@ -172,22 +172,26 @@ foam.CLASS({
...
-            b.key,
-            "'");
-
-          // Axioms have an option of wrapping a value for export.
-          // This could be used to bind a method to 'this', for example.
-          var e = a.exportAs ? a.exportAs(this, path.slice(1)) : this[path[0]];
-
-          m[b.exportName] = e;
+          if ( a ) {
+            foam.assert(
```
Comment (kgrgreer) [inline src/foam/core/ImportsExports.js:182]: Should the assert() be in the else?

File: src/foam/mlang/mlang.js [hunk 9]
```diff
@@ -4807,6 +4777,60 @@ foam.CLASS({
...
+        if ( result )
+          return getTrueExpr() != null ? getTrueExpr().f(obj) : null;
+        return getFalseExpr() != null ? getFalseExpr().f(obj) : null;
+      `
+    },
+    {
+      name: 'toString',
+      type: 'String',
+      javaCode: `
+        var sb = new StringBuilder();
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js]: FYI: If you do it all at once, then this is exactly the same as doing one list of string concatenations but with less code.

File: src/foam/core/ImportsExports.js (hunk 8 above)
Comment (jlhughes) [reply src/foam/core/ImportsExports.js:182]: moved

File: src/foam/mlang/mlang.js (hunk 9 above)
Comment (jlhughes) [reply src/foam/mlang/mlang.js]: updated

File: src/foam/mlang/mlang.js (hunk 5 above)
Comment (jlhughes) [reply src/foam/mlang/mlang.js]: updated

File: src/foam/nanos/auth/LastModifiedAware.js (hunk 4 above)
Comment (jlhughes) [reply src/foam/nanos/auth/LastModifiedAware.js]: reverted in final version.

File: src/foam/mlang/mlang.js (hunk 1 above)
Comment (jlhughes) [reply src/foam/mlang/mlang.js]: reverted in final version
